lib/notifications/notification_server.ex

```diff
@@ -0,0 +1,116 @@
+defmodule Notifications.NotificationServer do
+  use GenServer
+
+  alias Notifications.Notification
+  alias Realtime.BlockWaiver
+  alias Schedule.Block
+
+  require Logger
+
+  @type __MODULE__ :: {schedule_data_mod :: module()}
+
+  defstruct [:schedule_data_mod]
+
+  # Client
+
+  @spec default_name() :: GenServer.name()
+  def default_name(), do: Notifications.NotificationServer
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []) do
+    name = Keyword.get(opts, :name, __MODULE__)
+    schedule_data_mod = Keyword.get(opts, :schedule_data_mod, Schedule)
+    GenServer.start_link(__MODULE__, %{schedule_data_mod: schedule_data_mod}, name: name)
+  end
+
+  @spec new_block_waivers(BlockWaiver.block_waivers_by_block_key(), GenServer.server()) :: :ok
+  def new_block_waivers(new_waivers_by_block_key, server \\ default_name()) do
+    GenServer.cast(server, {:new_block_waivers, new_waivers_by_block_key})
+  end
+
+  # Server
+
+  @impl GenServer
+  def init(initial_state) do
+    schedule_data_mod = Map.fetch!(initial_state, :schedule_data_mod)
+    {:ok, %__MODULE__{schedule_data_mod: schedule_data_mod}}
+  end
+
+  @impl true
+  def handle_cast({:new_block_waivers, new_block_waivers_by_block_key}, state) do
+    new_notifications =
+      new_block_waivers_by_block_key
+      |> convert_new_block_waivers_to_notifications(state.schedule_data_mod)
+
+    if !Enum.empty?(new_notifications) do
+      Logger.warn(
+        "NotificationServer created new notifications new_notifications=#{
+          inspect(new_notifications)
+        }"
+      )
+    end
+
+    {:noreply, state}
+  end
+
+  @spec convert_new_block_waivers_to_notifications([BlockWaiver.t()], module()) :: [
+          Notification.t()
+        ]
+  defp convert_new_block_waivers_to_notifications(new_block_waivers, schedule_data_mod) do
+    new_block_waivers
+    |> Enum.flat_map(fn {block_key, block_waivers} ->
+      Enum.map(
+        block_waivers,
+        &convert_block_waiver_to_notification(block_key, &1, schedule_data_mod)
+      )
+    end)
+    |> Enum.filter(& &1)
+  end
+
+  @spec convert_block_waiver_to_notification(Block.key(), BlockWaiver.t(), module()) ::
+          Notification.t() | nil
+  defp convert_block_waiver_to_notification(
+         {block_id, service_id},
+         block_waiver,
+         schedule_data_mod
+       ) do
+    if reason = get_notification_reason(block_waiver) do
+      block = apply(schedule_data_mod, :block, [block_id, service_id])
```

Comment:
It turns out for the block waiver store, although it makes sense to `reassign_env` the function in question rather than sling a module around, the mock module is still handy for keeping a little state around.

Reply:
...scratch that, I think I see a good way around using the mock module.